api/path_node_userinput.go

```diff
@@ -115,3 +142,148 @@ func DeleteNodeUserInput(errorhandler DeviceErrorHandler, db *bolt.DB,
 	nodeUserInputUpdated := events.NewNodeUserInputMessage(events.UPDATE_NODE_USERINPUT, *chnagedSvcSpecs)
 	return false, []*events.NodeUserInputMessage{nodeUserInputUpdated}
 }
+
+var logString = func(v interface{}) string {
+	return fmt.Sprintf("path_node_userinput %v", v)
+}
+
+// Validate 1) service exist; 2) variables defined in the service; 3) values have correct type
+func ValidateUserInput(userInput policy.UserInput, getService exchange.ServiceHandler) (bool, error) {
+	glog.V(3).Infof(logString(fmt.Sprintf("Start validate userinput .... \n")))
+	serviceOrg := userInput.ServiceOrgid
+	serviceUrl := userInput.ServiceUrl
+	serviceVersionRange := userInput.ServiceVersionRange
+	serviceArch := userInput.ServiceArch
+
+	// The versionRange field is checked for valid characters by the Version_Expression_Factory, it has a very
+	// specific syntax and allows a subset of normally valid characters.
+
+	// Use a default sensor version that allows all version if not specified.
+	if &userInput.ServiceVersionRange == nil || userInput.ServiceVersionRange == "" {
+		def := "0.0.0"
+		serviceVersionRange = def
+	}
+
+	// Convert the sensor version to a version expression.
+	vExp, err := semanticversion.Version_Expression_Factory(serviceVersionRange)
+	var errorString string
+	if err != nil {
+		errorString = fmt.Sprintf("versionRange %v cannot be converted to a version expression, error %v", userInput.ServiceVersionRange, err)
+		return false, errors.New(errorString)
+	}
+
+	// service exist
+	var sdef *exchange.ServiceDefinition
+	sdef, _, err = getService(serviceUrl, serviceOrg, vExp.Get_expression(), serviceArch)
```

Comment:
If serviceArch is empty at this point, you should use cutil.ArchString() to get the node's arch so that you are retrieving the correct service for this node's arch. And now that I think about it, if serviceArch is not empty then it has to be the same as cutil.ArchString().

Reply:
Update:
1. Allows leave empty for serviceArch
2. Add a check for serviceArch = nodeArch if serviceArch is not empty, will return false if not match